editor/quick_open_dialog.cpp

```diff
@@ -0,0 +1,980 @@
+	_update_result_items(MIN(relevant_candidates, max_total_results), 0);
+}
+
+int QuickOpenResultContainer::_sort_candidates(const String &p_query) {
+	if (p_query.is_empty()) {
+		return 0;
+	}
+
+	// First, we assign a score based on the best found characteristic:
+	// 1. File substring match in file name
+	// 2. File substring match (positional)
+	// 3. File subsequence match
+	// 4. Directory substring match (positional)
+	// 5. Directory subsequence match
+	// 6. No match, file is not considered (-1).
+	int num_relevant_candidates = 0;
+	for (Candidate &c : candidates) {
+		const bool file_match = p_query.is_subsequence_ofn(c.file_name);
+		const bool dir_match = p_query.is_subsequence_ofn(c.file_directory);
+		if (!file_match && !dir_match) {
+			c.score = -1;
+			continue;
+		}
+
+		num_relevant_candidates++;
+
+		// File matches will always outrank directory matches.
+		float score = file_match ? 0.8f : 0.4f;
+
+		const String s = file_match ? c.file_name : c.file_directory;
+		int pos = s.findn(p_query);
+		if (pos > -1) {
+			// Positive bias for matches close to the beginning of the string.
+			score += 0.1f * (1.0f - ((float)pos / (float)s.length()));
+
+			if (file_match) {
+				// Positive bias for files if the match happened in the file name, not the extension.
+				int ext_pos = s.rfind(".");
+				if (ext_pos == -1 || ext_pos > pos) {
+					score += 0.1f;
+				}
+			}
+		}
+
+		c.score = score;
+	}
+
+	// Now we will sort the candidates based on score. However, within each
+	// characteristic, files can be tied in score. This is resolved by favoring:
+	// 1. Shorter file length.
+	// 2. Shorter directory length.
+	// 3. Lower alphabetic order.
+	struct CandidateComparator {
+		_FORCE_INLINE_ bool operator()(const Candidate &p_a, const Candidate &p_b) const {
+			if (abs(p_a.score - p_b.score) >= CMP_EPSILON) {
+				return p_a.score > p_b.score;
+			}
+
+			if (p_a.file_name.length() != p_b.file_name.length()) {
+				return p_a.file_name.length() < p_b.file_name.length();
+			}
+
+			if (p_a.file_directory.length() != p_b.file_directory.length()) {
+				return p_a.file_directory.length() < p_b.file_directory.length();
+			}
+
+			return p_a.file_name < p_b.file_name;
+		}
+	};
+	candidates.sort_custom<CandidateComparator>();
+
+	return num_relevant_candidates;
+}
+
+void QuickOpenResultContainer::_update_result_items(int p_new_visible_results_count, int p_new_selection_index) {
+	if (query.is_empty()) {
+		p_new_visible_results_count = (candidates.size() <= SHOW_ALL_FILES_THRESHOLD) ? candidates.size() : 0;
+	}
+
+	// Only need to update items that were not hidden in previous update.
+	int num_items_needing_updates = MAX(num_visible_results, p_new_visible_results_count);
+	num_visible_results = p_new_visible_results_count;
+
+	for (int i = 0; i < num_items_needing_updates; i++) {
+		QuickOpenResultItem *item = result_items[i];
+
+		if (i < num_visible_results) {
+			const Candidate &c = candidates.write[i];
```

Comment:
```suggestion
			const Candidate &c = candidates[i];
```
Const items don't need `write`.